kernel/src/device/tty/device.rs

```diff
@@ -25,12 +43,92 @@ impl Device for TtyDevice {
 
         terminal.open()
     }
+}
+
+/// Corresponds to `/dev/tty0` in the file system. This device represents the active virtual
+/// terminal.
+pub struct Tty0Device;
 
+impl Tty0Device {
+    fn active_vt(&self) -> &Arc<Tty<VtDriver>> {
+        // Currently there is only one virtual terminal `tty1`.
+        tty1_device()
+    }
+}
+
+impl Device for Tty0Device {
     fn type_(&self) -> DeviceType {
         DeviceType::Char
     }
 
     fn id(&self) -> DeviceId {
-        DeviceId::new(MajorId::new(5), MinorId::new(0))
+        DeviceId::new(MajorId::new(4), MinorId::new(0))
+    }
+
+    fn open(&self) -> Result<Box<dyn FileIo>> {
+        self.active_vt().open()
+    }
+}
+
+impl Terminal for Tty0Device {
+    fn job_control(&self) -> &JobControl {
+        self.active_vt().job_control()
+    }
+}
+
+/// Corresponds to `/dev/console` in the file system. This device represents a console to which
+/// system messages will be sent.
+pub struct SystemConsole {
+    inner: Arc<dyn Terminal>,
+}
+
+impl SystemConsole {
+    /// Returns the singleton instance of the console device.
+    pub fn singleton() -> &'static Arc<SystemConsole> {
+        static INSTANCE: Once<Arc<SystemConsole>> = Once::new();
+
+        INSTANCE.call_once(|| {
+            // TODO: Support specifying multiple TTY devices.
+            // e.g. "console=hvc0, console=tty0"
+            let console_name = boot_info()
+                .kernel_cmdline
+                .split_whitespace()
+                .find(|item| item.starts_with("console="))
+                .and_then(|value| value.split('=').nth(1))
```

Comment:
Use [`str::split_once`](https://doc.rust-lang.org/std/primitive.str.html#method.split_once) here?

It also seems that we have `kernel/src/kcmdline.rs`, but there is no reasonable way to access the parsed cmdline here?